### PR TITLE
rewrite PersistentNetworkMapService to use Hibernate

### PR DESCRIPTION
**Pull request by szymonsztuka**

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt [hunk 1]
```diff
@@ -26,9 +26,15 @@ class PersistentMap<K, V, E, EK> (
             concurrencyLevel = 8,
             loadFunction = { key -> Optional.ofNullable(loadValue(key)) },
             removalListener = ExplicitRemoval(toPersistentEntityKey, persistentEntityClass)
-    )
+    ).apply {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: Do we want to always load everything from the database on start, or should there be a flag?  Just wondering about any existing uses of `PersistentMap`...

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt [hunk 2]
```diff
@@ -90,28 +95,63 @@ class PersistentMap<K, V, E, EK> (
...
+     * @return true if added key was unique, otherwise false
+     */
+    fun addWithDuplicatesOverwritten(key: K, value: V) = //TODO rename to set
+            set(key, value,
+                    logWarning = false,
+                    store = { k: K, v: V -> merge(k, v) },
+                    override = { k: K, v:V -> override(k, v) }
+            )
+
+    private fun override(key: K, value: V) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: Is having a method and parameter called `override` a good idea?  It's quite confusing.  Can you come up with another name?

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt (hunk 1 above)
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: Do other services (that your're in the process of converting) need full map behaviour or is it just the network map?

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt [hunk 3]
```diff
@@ -46,38 +52,37 @@ class PersistentMap<K, V, E, EK> (
...
         return cache.get(key).orElse(null)
     }
 
     fun all(): Sequence<Pair<K, V>> {
-        return cache.asMap().map { entry -> Pair(entry.key as K, entry.value as V) }.asSequence()
+        return cache.asMap().map { entry -> Pair(entry.key as K, entry.value.get()) }.asSequence()
     }
 
-    private tailrec fun set(key: K, value: V, logWarning: Boolean = true, store: (K,V) -> V?): Boolean {
+    override val size = all().count()
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: This seems very inefficient - converting everything to a sequence and then counting. Can you not just ask for the size of the cache?

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt (hunk 1 above)
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: As discussed.... looks like there's sufficient use to justify, and all that use is "in memory", so leave as is.

File: node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt (hunk 2 above)
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/utilities/PersistentMap.kt]: Changed to replace (terminology from java.util.Map

**Pull request merged**